File: packages/agents/src/tracing.py
"""
Distributed tracing configuration using OpenTelemetry.

This module provides instrumentation for tracing requests across agents
and external services.
"""
from typing import Optional, Dict, Any, Callable
from functools import wraps
import os

from opentelemetry import trace
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor, ConsoleSpanExporter
from opentelemetry.sdk.resources import Resource, SERVICE_NAME, SERVICE_VERSION
from opentelemetry.exporter.otlp.proto.grpc.trace_exporter import OTLPSpanExporter
from opentelemetry.instrumentation.requests import RequestsInstrumentor
from opentelemetry.instrumentation.redis import RedisInstrumentor
from opentelemetry.instrumentation.pymongo import PymongoInstrumentor
from opentelemetry.trace import Status, StatusCode, Span
from opentelemetry.trace.propagation.tracecontext import TraceContextTextMapPropagator


# Global tracer instance
_tracer: Optional[trace.Tracer] = None


def setup_tracing(
    service_name: str = "chimera-agent",
    service_version: str = "1.0.0",
    jaeger_endpoint: Optional[str] = None,
    enable_console_export: bool = False,
) -> trace.Tracer:
    """
    Set up OpenTelemetry tracing for the application.
    
    Args:
        service_name: Name of the service for trace identification
        service_version: Version of the service
        jaeger_endpoint: Jaeger collector endpoint (e.g., "http://localhost:4317")
        enable_console_export: Whether to export traces to console for debugging
    
    Returns:
        Configured tracer instance
    """
    global _tracer
    
    # Create resource with service information
    resource = Resource.create({
        SERVICE_NAME: service_name,
        SERVICE_VERSION: service_version,
    })
    
    # Create tracer provider
    provider = TracerProvider(resource=resource)
    
    # Add span processors
    if jaeger_endpoint or os.getenv("OTEL_EXPORTER_OTLP_ENDPOINT"):
        # Use OTLP exporter for Jaeger
        otlp_exporter = OTLPSpanExporter(
            endpoint=jaeger_endpoint or os.getenv("OTEL_EXPORTER_OTLP_ENDPOINT"),
            insecure=True,  # Use insecure for local development
        )
        provider.add_span_processor(BatchSpanProcessor(otlp_exporter))
    
    if enable_console_export:
        # Add console exporter for debugging
        console_exporter = ConsoleSpanExporter()
        provider.add_span_processor(BatchSpanProcessor(console_exporter))
    
    # Set the tracer provider
    trace.set_tracer_provider(provider)
    
    # Instrument libraries
    RequestsInstrumentor().instrument()
    RedisInstrumentor().instrument()
    PymongoInstrumentor().instrument()
    
    # Get tracer
    _tracer = trace.get_tracer(__name__)
    
    return _tracer


def get_tracer() -> trace.Tracer:
    """
    Get the global tracer instance.
    
    Returns:
        Tracer instance
    
    Raises:
        RuntimeError: If tracing has not been set up
    """
    if _tracer is None:
        raise RuntimeError("Tracing not initialized. Call setup_tracing() first.")
    return _tracer


def create_span(
    name: str,
    attributes: Optional[Dict[str, Any]] = None,
    kind: trace.SpanKind = trace.SpanKind.INTERNAL,
) -> Span:
    """
    Create a new span.
    
    Args:
        name: Name of the span
        attributes: Optional attributes to add to the span
        kind: Type of span (INTERNAL, SERVER, CLIENT, PRODUCER, CONSUMER)
    
    Returns:
        Span instance
    """
    tracer = get_tracer()
    span = tracer.start_span(name, kind=kind)
    
    if attributes:
        for key, value in attributes.items():
            span.set_attribute(key, value)
    
    return span


def trace_function(
    name: Optional[str] = None,
    attributes: Optional[Dict[str, Any]] = None,
) -> Callable:
    """
    Decorator to trace a function execution.
    
    Args:
        name: Optional custom name for the span (defaults to function name)
        attributes: Optional attributes to add to the span
    
    Returns:
        Decorated function
    """
    def decorator(func: Callable) -> Callable:
        @wraps(func)
        def wrapper(*args, **kwargs):
            tracer = get_tracer()
            span_name = name or f"{func.__module__}.{func.__name__}"
            
            with tracer.start_as_current_span(span_name) as span:
                # Add attributes
                if attributes:
                    for key, value in attributes.items():
                        span.set_attribute(key, value)
                
                # Add function info
                span.set_attribute("function.name", func.__name__)
                span.set_attribute("function.module", func.__module__)
                
                try:
                    result = func(*args, **kwargs)
                    span.set_status(Status(StatusCode.OK))
                    return result
                except Exception as e:
                    span.set_status(Status(StatusCode.ERROR, str(e)))
                    span.record_exception(e)
                    raise
        
        return wrapper
    return decorator


def trace_agent_operation(operation: str) -> Callable:
    """
    Decorator to trace agent operations.
    
    Args:
        operation: Name of the operation (e.g., "process_query", "fetch_data")
    
    Returns:
        Decorated function
    """
    return trace_function(
        name=f"agent.{operation}",
        attributes={"operation.type": "agent"}
    )


def inject_trace_context(carrier: Dict[str, str]) -> None:
    """
    Inject trace context into a carrier (e.g., message headers).
    
    Args:
        carrier: Dictionary to inject trace context into
    """
    propagator = TraceContextTextMapPropagator()
    propagator.inject(carrier)


def extract_trace_context(carrier: Dict[str, str]) -> None:
    """
    Extract trace context from a carrier and set as current context.
    
    Args:
        carrier: Dictionary containing trace context
    """
    propagator = TraceContextTextMapPropagator()
    context = propagator.extract(carrier)
    trace.set_span_in_context(trace.get_current_span(), context)


def add_span_event(name: str, attributes: Optional[Dict[str, Any]] = None) -> None:
    """
    Add an event to the current span.
    
    Args:
        name: Name of the event
        attributes: Optional attributes for the event
    """
    span = trace.get_current_span()
    if span:
        span.add_event(name, attributes or {})


def set_span_attribute(key: str, value: Any) -> None:
    """
    Set an attribute on the current span.
    
    Args:
        key: Attribute key
        value: Attribute value
    """
    span = trace.get_current_span()
    if span:
        span.set_attribute(key, value)


def set_span_error(error: Exception) -> None:
    """
    Mark the current span as error and record the exception.
    
    Args:
        error: Exception that occurred
    """
    span = trace.get_current_span()
    if span:
        span.set_status(Status(StatusCode.ERROR, str(error)))
        span.record_exception(error)


# Context manager for manual span creation
class traced_span:
    """
    Context manager for creating traced spans.
    
    Usage:
        with traced_span("operation_name", attributes={"key": "value"}):
            # Your code here
            pass
    """
    
    def __init__(
        self,
        name: str,
        attributes: Optional[Dict[str, Any]] = None,
        kind: trace.SpanKind = trace.SpanKind.INTERNAL,
    ):
        self.name = name
        self.attributes = attributes or {}
        self.kind = kind
        self.span: Optional[Span] = None
    
    def __enter__(self) -> Span:
        tracer = get_tracer()
        self.span = tracer.start_span(self.name, kind=self.kind)
        
        for key, value in self.attributes.items():
            self.span.set_attribute(key, value)
        
        return self.span
    
    def __exit__(self, exc_type, exc_val, exc_tb):
        if self.span:
            if exc_type is not None:
                self.span.set_status(Status(StatusCode.ERROR, str(exc_val)))
                self.span.record_exception(exc_val)
            else:
                self.span.set_status(Status(StatusCode.OK))
            
            self.span.end()


# Helper functions for common tracing patterns
def trace_message_processing(
    message_type: str,
    correlation_id: str,
) -> traced_span:
    """
    Create a span for message processing.
    
    Args:
        message_type: Type of message being processed
        correlation_id: Correlation ID for the message
    
    Returns:
        Context manager for the span
    """
    return traced_span(
        f"message.process.{message_type}",
        attributes={
            "message.type": message_type,
            "correlation.id": correlation_id,
        },
        kind=trace.SpanKind.CONSUMER,
    )


def trace_external_call(
    service: str,
    operation: str,
) -> traced_span:
    """
    Create a span for external service calls.
    
    Args:
        service: Name of the external service
        operation: Operation being performed
    
    Returns:
        Context manager for the span
    """
    return traced_span(
        f"external.{service}.{operation}",
        attributes={
            "service.name": service,
            "operation": operation,
        },
        kind=trace.SpanKind.CLIENT,
    )


def trace_database_operation(
    database: str,
    operation: str,
    collection: Optional[str] = None,
) -> traced_span:
    """
    Create a span for database operations.
    
    Args:
        database: Database name
        operation: Operation being performed (e.g., "find", "insert")
        collection: Optional collection/table name
    
    Returns:
        Context manager for the span
    """
    attributes = {
        "db.system": database,
        "db.operation": operation,
    }
    
    if collection:
        attributes["db.collection"] = collection
    
    return traced_span(
        f"db.{database}.{operation}",
        attributes=attributes,
        kind=trace.SpanKind.CLIENT,
    )

File: packages/api/src/tracing.ts
/**
 * Distributed tracing configuration using OpenTelemetry.
 * 
 * This module provides instrumentation for tracing requests across services
 * and external dependencies.
 */
import { NodeSDK } from '@opentelemetry/sdk-node';
import { getNodeAutoInstrumentations } from '@opentelemetry/auto-instrumentations-node';
import { Resource } from '@opentelemetry/resources';
import { SemanticResourceAttributes } from '@opentelemetry/semantic-conventions';
import { OTLPTraceExporter } from '@opentelemetry/exporter-trace-otlp-grpc';
import { BatchSpanProcessor, ConsoleSpanExporter } from '@opentelemetry/sdk-trace-base';
import { trace, context, Span, SpanStatusCode, SpanKind } from '@opentelemetry/api';
import { W3CTraceContextPropagator } from '@opentelemetry/core';

let sdk: NodeSDK | null = null;
let tracer: ReturnType<typeof trace.getTracer> | null = null;

/**
 * Initialize OpenTelemetry tracing
 */
export function setupTracing(
  serviceName: string = 'chimera-api',
  serviceVersion: string = '1.0.0',
  jaegerEndpoint?: string,
  enableConsoleExport: boolean = false,
): void {
  const resource = Resource.default().merge(
    new Resource({
      [SemanticResourceAttributes.SERVICE_NAME]: serviceName,
      [SemanticResourceAttributes.SERVICE_VERSION]: serviceVersion,
    }),
  );

  const traceExporter = jaegerEndpoint || process.env.OTEL_EXPORTER_OTLP_ENDPOINT
    ? new OTLPTraceExporter({
        url: jaegerEndpoint || process.env.OTEL_EXPORTER_OTLP_ENDPOINT,
      })
    : undefined;

  const spanProcessors = [];
  
  if (traceExporter) {
    spanProcessors.push(new BatchSpanProcessor(traceExporter));
  }
  
  if (enableConsoleExport) {
    spanProcessors.push(new BatchSpanProcessor(new ConsoleSpanExporter()));
  }

  sdk = new NodeSDK({
    resource,
    spanProcessor: spanProcessors.length > 0 ? spanProcessors[0] : undefined,
    instrumentations: [
      getNodeAutoInstrumentations({
        '@opentelemetry/instrumentation-fs': {
          enabled: false, // Disable fs instrumentation to reduce noise
        },
      }),
    ],
    textMapPropagator: new W3CTraceContextPropagator(),
  });

  sdk.start();
  tracer = trace.getTracer(serviceName, serviceVersion);
}

/**
 * Shutdown tracing gracefully
 */
export async function shutdownTracing(): Promise<void> {
  if (sdk) {
    await sdk.shutdown();
  }
}

/**
 * Get the tracer instance
 */
export function getTracer(): ReturnType<typeof trace.getTracer> {
  if (!tracer) {
    throw new Error('Tracing not initialized. Call setupTracing() first.');
  }
  return tracer;
}

/**
 * Create a new span
 */
export function createSpan(
  name: string,
  attributes?: Record<string, string | number | boolean>,
  kind: SpanKind = SpanKind.INTERNAL,
): Span {
  const currentTracer = getTracer();
  const span = currentTracer.startSpan(name, {
    kind,
    attributes,
  });
  return span;
}

/**
 * Decorator to trace a function
 */
export function traceFunction(
  name?: string,
  attributes?: Record<string, string | number | boolean>,
) {
  return function (
    target: any,
    propertyKey: string,
    descriptor: PropertyDescriptor,
  ) {
    const originalMethod = descriptor.value;
    const spanName = name || `${target.constructor.name}.${propertyKey}`;

    descriptor.value = async function (...args: any[]) {
      const currentTracer = getTracer();
      return currentTracer.startActiveSpan(spanName, async (span) => {
        try {
          if (attributes) {
            Object.entries(attributes).forEach(([key, value]) => {
              span.setAttribute(key, value);
            });
          }

          span.setAttribute('function.name', propertyKey);
          span.setAttribute('function.class', target.constructor.name);

          const result = await originalMethod.apply(this, args);
          span.setStatus({ code: SpanStatusCode.OK });
          return result;
        } catch (error) {
          span.setStatus({
            code: SpanStatusCode.ERROR,
            message: error instanceof Error ? error.message : String(error),
          });
          span.recordException(error as Error);
          throw error;
        } finally {
          span.end();
        }
      });
    };

    return descriptor;
  };
}

/**
 * Trace an async function
 */
export async function traceAsync<T>(
  name: string,
  fn: (span: Span) => Promise<T>,
  attributes?: Record<string, string | number | boolean>,
  kind: SpanKind = SpanKind.INTERNAL,
): Promise<T> {
  const currentTracer = getTracer();
  return currentTracer.startActiveSpan(
    name,
    { kind, attributes },
    async (span) => {
      try {
        const result = await fn(span);
        span.setStatus({ code: SpanStatusCode.OK });
        return result;
      } catch (error) {
        span.setStatus({
          code: SpanStatusCode.ERROR,
          message: error instanceof Error ? error.message : String(error),
        });
        span.recordException(error as Error);
        throw error;
      } finally {
        span.end();
      }
    },
  );
}

/**
 * Trace a synchronous function
 */
export function traceSync<T>(
  name: string,
  fn: (span: Span) => T,
  attributes?: Record<string, string | number | boolean>,
  kind: SpanKind = SpanKind.INTERNAL,
): T {
  const currentTracer = getTracer();
  return currentTracer.startActiveSpan(
    name,
    { kind, attributes },
    (span) => {
      try {
        const result = fn(span);
        span.setStatus({ code: SpanStatusCode.OK });
        return result;
      } catch (error) {
        span.setStatus({
          code: SpanStatusCode.ERROR,
          message: error instanceof Error ? error.message : String(error),
        });
        span.recordException(error as Error);
        throw error;
      } finally {
        span.end();
      }
    },
  );
}

/**
 * Inject trace context into carrier (e.g., HTTP headers)
 */
export function injectTraceContext(carrier: Record<string, string>): void {
  const propagator = new W3CTraceContextPropagator();
  propagator.inject(context.active(), carrier, {
    set: (carrier, key, value) => {
      carrier[key] = value;
    },
  });
}

/**
 * Extract trace context from carrier
 */
export function extractTraceContext(carrier: Record<string, string>): void {
  const propagator = new W3CTraceContextPropagator();
  const extractedContext = propagator.extract(context.active(), carrier, {
    get: (carrier, key) => carrier[key],
    keys: (carrier) => Object.keys(carrier),
  });
  context.with(extractedContext, () => {});
}

/**
 * Add event to current span
 */
export function addSpanEvent(
  name: string,
  attributes?: Record<string, string | number | boolean>,
): void {
  const span = trace.getActiveSpan();
  if (span) {
    span.addEvent(name, attributes);
  }
}

/**
 * Set attribute on current span
 */
export function setSpanAttribute(
  key: string,
  value: string | number | boolean,
): void {
  const span = trace.getActiveSpan();
  if (span) {
    span.setAttribute(key, value);
  }
}

/**
 * Mark current span as error
 */
export function setSpanError(error: Error): void {
  const span = trace.getActiveSpan();
  if (span) {
    span.setStatus({
      code: SpanStatusCode.ERROR,
      message: error.message,
    });
    span.recordException(error);
  }
}

/**
 * Helper functions for common tracing patterns
 */

export function traceHttpRequest(
  method: string,
  path: string,
  statusCode?: number,
): Span {
  return createSpan(
    `HTTP ${method} ${path}`,
    {
      'http.method': method,
      'http.route': path,
      ...(statusCode && { 'http.status_code': statusCode }),
    },
    SpanKind.SERVER,
  );
}

export function traceMessageProcessing(
  messageType: string,
  correlationId: string,
): Span {
  return createSpan(
    `message.process.${messageType}`,
    {
      'message.type': messageType,
      'correlation.id': correlationId,
    },
    SpanKind.CONSUMER,
  );
}

export function traceExternalCall(
  service: string,
  operation: string,
): Span {
  return createSpan(
    `external.${service}.${operation}`,
    {
      'service.name': service,
      'operation': operation,
    },
    SpanKind.CLIENT,
  );
}

export function traceDatabaseOperation(
  database: string,
  operation: string,
  collection?: string,
): Span {
  const attributes: Record<string, string> = {
    'db.system': database,
    'db.operation': operation,
  };

  if (collection) {
    attributes['db.collection'] = collection;
  }

  return createSpan(
    `db.${database}.${operation}`,
    attributes,
    SpanKind.CLIENT,
  );
}

/**
 * Middleware to trace HTTP requests
 */
export function tracingMiddleware() {
  return (req: any, res: any, next: any) => {
    const currentTracer = getTracer();
    const span = currentTracer.startSpan(`HTTP ${req.method} ${req.route?.path || req.path}`, {
      kind: SpanKind.SERVER,
      attributes: {
        'http.method': req.method,
        'http.url': req.url,
        'http.route': req.route?.path || req.path,
        'correlation.id': req.correlationId,
      },
    });

    // Store span in request for later access
    req.span = span;

    res.on('finish', () => {
      span.setAttribute('http.status_code', res.statusCode);
      if (res.statusCode >= 400) {
        span.setStatus({
          code: SpanStatusCode.ERROR,
          message: `HTTP ${res.statusCode}`,
        });
      } else {
        span.setStatus({ code: SpanStatusCode.OK });
      }
      span.end();
    });

    next();
  };
}
